src/NUnitEngine/nunit.engine/Agent/AgentClient.cs

```diff
@@ -0,0 +1,174 @@
+﻿// ***********************************************************************
+// Copyright (c) 2019 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+#if !NETSTANDARD1_6 // Dependency on BinaryFormatter
+using System;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+using System.Threading;
+
+namespace NUnit.Engine.Agent
+{
+    internal sealed class AgentClient : IDisposable
+    {
+        private readonly Stream _stream;
+        private readonly BinaryReader _reader;
+        private readonly BinaryWriter _writer;
+        private readonly BinaryFormatter _formatter = new BinaryFormatter();
+        private volatile StopRunState _stopRun;
+
+        private enum StopRunState
+        {
+            None = 0,
+            Stop,
+            Force
+        }
+
+        public AgentClient(Stream stream)
+        {
+            _stream = stream ?? throw new ArgumentNullException(nameof(stream));
+            _reader = new BinaryReader(stream);
+            _writer = new BinaryWriter(stream);
+        }
+
+        public void Dispose()
+        {
+            _stream.Dispose();
+        }
+
+        public void Connect(TestPackage package)
+        {
+            if (package == null) throw new ArgumentNullException(nameof(package));
+
+            const byte version = 1;
+            _writer.Write(version);
+            _formatter.Serialize(_stream, package);
+            _stream.Flush();
+        }
+
+        public TestEngineResult Load()
+        {
+            WriteCommandType(AgentCommandType.Load);
+            _stream.Flush();
+
+            return ReadTestEngineResult();
```

Comment:
This shows why opening the port on the agent is easier, it fits well with the current architecture of the engine where we send commands to the agent and wait for the response. I'd done a much smaller spike on this before but with the engine opening the port and ended up jumping through hoops because it was somewhat backwards.